sandbox/host/src/sandbox/wasmi_backend.rs

```diff
@@ -18,393 +18,534 @@
-        BackendInstanceBundle, GuestEnvironment, GuestExternals, GuestFuncIndex, Imports,
-        InstantiationError, Memory, SandboxInstance, SupervisorContext,
+        BackendInstanceBundle, GuestEnvironment, InstantiationError, Memory, SandboxInstance,
+        SupervisorContext,
     },
+    store_refcell,
     util::MemoryTransfer,
 };
 
+use super::SupervisorFuncIndex;
+
+type Store = wasmi::Store<Option<FuncEnv>>;
+pub type StoreRefCell = store_refcell::StoreRefCell<Store>;
+
 environmental::environmental!(SupervisorContextStore: trait SupervisorContext);
 
-#[derive(Debug)]
-struct CustomHostError(String);
+pub struct FuncEnv {
+    store: Weak<StoreRefCell>,
+    gas_global: wasmi::Global,
+}
 
-impl fmt::Display for CustomHostError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "HostError: {}", self.0)
+impl FuncEnv {
+    pub fn new(store: Weak<StoreRefCell>, gas_global: wasmi::Global) -> Self {
+        Self { store, gas_global }
     }
 }
 
-impl sandbox_wasmi::HostError for CustomHostError {}
-
 /// Construct trap error from specified message
-fn trap(msg: &'static str) -> Trap {
-    Trap::host(CustomHostError(msg.into()))
+fn host_trap(msg: impl Into<error::Error>) -> wasmi::Error {
+    wasmi::Error::host(msg.into())
+}
+
+fn into_wasmi_val(value: Value) -> wasmi::Val {
+    match value {
+        Value::I32(val) => wasmi::Val::I32(val),
+        Value::I64(val) => wasmi::Val::I64(val),
+        Value::F32(val) => wasmi::Val::F32(wasmi::core::F32::from_bits(val)),
+        Value::F64(val) => wasmi::Val::F64(wasmi::core::F64::from_bits(val)),
+    }
 }
 
-impl ImportResolver for Imports {
-    fn resolve_func(
-        &self,
-        module_name: &str,
-        field_name: &str,
-        signature: &sandbox_wasmi::Signature,
-    ) -> std::result::Result<sandbox_wasmi::FuncRef, sandbox_wasmi::Error> {
-        let idx = self.func_by_name(module_name, field_name).ok_or_else(|| {
-            sandbox_wasmi::Error::Instantiation(format!(
-                "Export {}:{} not found",
-                module_name, field_name
-            ))
-        })?;
-
-        Ok(sandbox_wasmi::FuncInstance::alloc_host(
-            signature.clone(),
-            idx.0,
-        ))
+fn into_wasmi_result(value: ReturnValue) -> Vec<wasmi::Val> {
+    match value {
+        ReturnValue::Value(v) => vec![into_wasmi_val(v)],
+        ReturnValue::Unit => vec![],
     }
+}
+
+fn into_value(value: &wasmi::Val) -> Option<Value> {
+    match value {
+        wasmi::Val::I32(val) => Some(Value::I32(*val)),
+        wasmi::Val::I64(val) => Some(Value::I64(*val)),
+        wasmi::Val::F32(val) => Some(Value::F32(u32::from(*val))),
+        wasmi::Val::F64(val) => Some(Value::F64(u64::from(*val))),
+        _ => None,
+    }
+}
+
+/// Wasmi specific context
+pub struct Backend {
+    store: Rc<StoreRefCell>,
+    // Allocation should be dropped right after the store is dropped
+    allocations: Vec<Allocation>,
```

Comment:
Invariant can be violated. Allocations can be dropped before store is dropped because of `Rc` thus lifetime of store can be longer than lifetime of allocations. You need to keep `StoreRefCell` and `Vec<Allocation>` both in the same `Rc`.

Reply:
Added check in `Drop`